module/zstd/zfs_zstd.c

```diff
@@ -545,189 +609,71 @@ int
 zfs_zstd_decompress(void *s_start, void *d_start, size_t s_len, size_t d_len,
     int level __maybe_unused)
 {
-
 	return (zfs_zstd_decompress_level(s_start, d_start, s_len, d_len,
 	    NULL));
 }
 
-/* Allocator for zstd compression context using mempool_allocator */
+/* supplied as custom allocator() to zstd code */
 static void *
-zstd_alloc(void *opaque __maybe_unused, size_t size)
+zstd_alloc_cb(void *opaque __maybe_unused, size_t size)
 {
-	size_t nbytes = sizeof (struct zstd_kmem) + size;
-	struct zstd_kmem *z = NULL;
-
-	z = (struct zstd_kmem *)zstd_mempool_alloc(zstd_mempool_cctx, nbytes);
+	int const try_harder = (opaque != NULL);
+	struct zstd_kmem_hdr *z;
+	size_t nbytes = sizeof (*z) + size;
 
+	z = vmem_alloc(nbytes, KM_NOSLEEP);
```

Comment:
This concerns me, we probably should not even allow using `KM_NOSLEEP` with `vmem_alloc()` since this will eventually call `__vmalloc( , GFP_ATOMIC)` and my understanding is the atomic flag will be ignored and it may sleep anyway.

Always using `vmem_alloc(nbytes, KM_SLEEP)` is probably a better option and let it handle the retry fallback logic.  On Linux for buffer sizes <= 1M (`spl_kmem_alloc_max`) it will prefer `kvmalloc()` which does what you want, and for larger buffers it'll default to `vmalloc()`.  We should just make sure that with 16M record sizes performance is comparable.  On FreeBSD this all maps to `zfs_kmem_alloc()` which I believe should be fine.  Dropping the `zstd_stat_alloc_fail` seems like it would be fine.

Speaking of FreeBSD I also noticed both FreeBSD builds crashed why testing this PR so something definitely isn't quite right on the FreeBSD side.

http://build.zfsonlinux.org/builders/FreeBSD%20stable%2F13%20amd64%20%28TEST%29/builds/386

Reply:
KM_SLEEP can lead to deadlocks if memory is exausted. that was a experience i had in linux while testing a long time ago. not a good option for larger blocks. it will deadlock the system until you push the reset button.